R/primerTree.R

```diff
@@ -264,3 +267,118 @@ labeled_quantile = function(x, labels, ...){
+#' there is only one species is omitted, as the number of nucleotide
+#' differences will always be 0.
+#' 
+#' @examples
+#' \dontrun{
+#' calc_rank_dist_ave(mammals_16S)
+#' 
+#' calc_rank_dist_ave(bryophytes_trnL)
+#' 
+#' # Note that the differences between the results from these two primers
+#' # the mean nucleotide differences is much higher for the mammal primers
+#' # than the byrophyte primers. This suggests that the mammal primers have
+#' # better resolution to distinguish individual species.
+#' }
+#' @export
+
+# using tree data format info from http://www.phytools.org/eqg/Exercise_3.2/ 
+calc_rank_dist_ave <- function(x, ranks = common_ranks, ...) {
+  usedRanks <- grep("species", ranks, invert = T, value = T)
+  rankDistMean <- data.frame(matrix(nrow=1,ncol=0))
+  for(rank in usedRanks) {
+    
+    # Raw taxonomy data
+    taxa <- as.data.frame(x$taxonomy)
+    
+    # Randomize the order of the taxa data frame for the next step
+    taxa <- taxa[ sample(1:nrow(taxa), size = nrow(taxa), replace = F) ,]
+    rownames(taxa) <- taxa$gi
+    
+    # Pick random example per species and add back in the taxa info
+    uniqueFactors <- as.data.frame(unique( taxa$species ))
+    colnames(uniqueFactors) <- "species"
+    uniqueFactors <- join(uniqueFactors, taxa, type = "left",match="first", by = "species")
+    uniqueFactors <- uniqueFactors[, colnames(uniqueFactors) %in% c("gi","species",rank)]
+    
+    # Drop any row in (rank) where there is only one species represented
+    # Any instance of this leads to a distance within that rank of 0, skewing the results downward
+    counts <- as.data.frame(table(uniqueFactors[[rank]]))
+    colnames(counts) <- c(rank,"count")
+    uniqueFactors <- join(uniqueFactors, counts, by = rank)
+    uniqueFactors <- subset(uniqueFactors, count > 1)
+    
+    # Get sequences for randomly selected species
+    seqs <- x$sequence
+    seqs <- seqs[names(seqs) %in% uniqueFactors$gi ]
+    
+    # Align and calculate pairwise distances and convert dists to dataframe
+    align <- clustalo(seqs)
+    dists <- as.data.frame(as.matrix(dist.dna(align, model="N", pairwise.deletion=T)))
+    dists$gi <- row.names(dists)
+    
+    # Melt the dists dataframe so I can drop any distance that isn't within the (rank)
+    melted <- melt(dists,id="gi",variable.name = "gi2", value.name = "dist")
+    
+    # Replace the rank1 gi with the rank1 taxa
+    #melted2 <- join(melted, uniqueFactors[,colnames(uniqueFactors) %in% c("gi", rank)],by = "gi")
+    melted <- join(melted, uniqueFactors,by = "gi")
+    melted$rank1 <- melted[[rank]]
+    
+    # Drop all columns except the three needed so the next join doesn't get messed up
+    melted <- melted[,colnames(melted) %in% c("gi2", "dist", "rank1", "species")]
+    colnames(melted)[1] <- "gi"
+    
+    # Replace the rank2 gi with the rank2 taxa
+    melted <- join(melted, uniqueFactors, by = "gi")
+    melted$rank2 <- melted[[rank]]
+    
```

Comment:
Yeah, it's a bit complex. I'm essentially pulling out specific data points from the distance matrix and calculating the average of those values. The data I want are the values where the taxonomic group is the same for the two sequences compared (ie: both genus Canis). Since the rows/columns are labeled by gi, I start by melting the matrix to long form and then replace the the rowname/colname gis with the taxon name so I can keep only the comparisons within a taxa. That's why there are two merge operations (I replace row gi with taxa, then replace column gi with taxa). I'm also randomly sampling one sequence per species to avoid situations where the distance between all the 500 human sequences are ~0, which drags the average way down artificially. Maybe there is a simpler way to do this, but this was the most efficient way I could think of to pull out the relevant data points to calculate the mean nucleotide difference.